aten/src/ATen/core/ivalue.h

```diff
@@ -609,17 +609,26 @@ inline optional<T> IValue::toOptional() {
 }
 
 inline bool IValue::isSameIdentity(IValue& rhs) {
-  // We choose to not use memcmp for payload check due to potenntial random padding characters on union type
+  // We choose to not use memcmp for payload check due to potential random padding characters on union type
 
   // Semantics:
   // 1. None is None, False is False, and True is True are all true
-  // 2. If it is a reference type (i.e. is_intrusive_ptr), then is is True when the pointed-to object is the same.
-  // 3. False for all other comparisons.
+  // 2. If it is a tensor type, we need to take undefined tensor into account
+  // 3. Undefined_tensor is None should be true
+  // 4. If it is a reference type (i.e. is_intrusive_ptr), then is is True when the pointed-to object is the same.
+  // 5. False for all other comparisons.
+  std::cout<<"this tag kind: " << this->tagKind() << " rhs tag kind: " << rhs.tagKind() << std::endl;
   if (this->isNone() && rhs.isNone()) {
     return true;
   } else if (this->isBool() && rhs.isBool()) {
     // for bool type, do equality check
     return this->toBool() == rhs.toBool();
+  } else if (this->isTensor() && rhs.isTensor()) {
+    // for tensor type, just check the as_intrusive_ptr since is_intrusive_ptr is false for undefined tensor
+    return this->payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
+  } else if (this->isTensor() && rhs.isNone()) {
```

Comment:
Do we not need to check for the case where this.isNone && rhs.isTensor()  ?

Reply:
Something in Python like: `None is bias`? which I think it may never gonna happen.. but we can add it in case

Reply:
It could occur with an optional variable